### PR TITLE
(refactor) update to compile with a recent rustc

### PR DESCRIPTION
**Pull request by hugoduncan**

### CODE REVIEW

File: examples/get_request.rs [hunk 1]
```diff
@@ -1,21 +1,22 @@
...
     // Extract the decoded data as hashmap, using the UrlEncodedQuery plugin.
     match req.get_ref::<UrlEncodedQuery>() {
-        Some(hashmap) => println!("Parsed GET request query string:\n {}", hashmap),
-        None => println!("Error, no query string found")
+        Ok(hashmap) => println!("Parsed GET request query string:\n {:?}", hashmap),
+        Err(x) => println!("Error, no query string found: {:?}", x)
     };
 
-    Ok(Response::new().set(Status(status::Ok)).set(Body("Hello!")))
+    Ok(Response::new().set((status::Ok,"Hello!")))
```
Comment (reem) [inline examples/get_request.rs]: Just use `Response::with((status::Ok, "Hello!"))`.

File: examples/post_request.rs [hunk 2]
```diff
@@ -1,19 +1,20 @@
+#![feature(io)]
+
 // This example is almost the same as get_request.rs, but uses UrlEncodedBody
 extern crate iron;
 extern crate urlencoded;
 
 use iron::{Iron, Request, Response, IronResult, Plugin, Set, status};
```
Comment (reem) [inline examples/post_request.rs]: use `iron::prelude::*`

File: examples/post_request.rs [hunk 3]
```diff
@@ -1,19 +1,20 @@
...
 fn log_post_data(req: &mut Request) -> IronResult<Response> {
     match req.get_ref::<UrlEncodedBody>() {
-        Some(ref hashmap) => println!("Parsed POST request body:\n {}", hashmap),
-        None => println!("Error, no body found.")
+        Ok(ref hashmap) => println!("Parsed POST request body:\n {:?}", hashmap),
+        Err(x) => println!("Error, no body found: {:?}", x)
     };
 
-    Ok(Response::new().set(Status(status::Ok)).set(Body("Hello!")))
+    Ok(Response::new().set((status::Ok,"Hello!")))
```
Comment (reem) [inline examples/post_request.rs]: Response::with again

File: src/lib.rs [hunk 4]
```diff
@@ -33,48 +30,68 @@ pub struct UrlEncodedQuery;
...
 }
 
-
-impl PluginFor<Request, HashMap<String, Vec<String>>> for UrlEncodedBody {
-    fn eval(req: &mut Request, _: Phantom<UrlEncodedBody>) -> Option<QueryMap> {
-        str::from_utf8(req.body.as_slice()).and_then(create_param_hashmap)
+impl<'a> Plugin<Request<'a>> for UrlEncodedBody {
+    type Error = ErrorType;
+    fn eval(req: &mut Request) -> Result<QueryMap, ErrorType> {
+        str::from_utf8(&req.body.read_to_end().unwrap())
```
Comment (reem) [inline src/lib.rs]: This should use the body reader plugin

File: src/lib.rs [hunk 5]
```diff
@@ -33,48 +30,68 @@ pub struct UrlEncodedQuery;
 /// Use it like this: `req.get_ref::<UrlEncodedBody>()`
 pub struct UrlEncodedBody;
 
+#[derive(Clone, Debug, PartialEq)]
+pub enum ErrorType{
```
Comment (reem) [inline src/lib.rs]: Amazing. Can you implement `std::error::Error` for it too?

Comment (reem) [inline src/lib.rs]: Also a different name, like `UrlDecodingError`, would be better. If you can think of something that fits better, feel free.

File: src/lib.rs [hunk 6]
```diff
@@ -33,48 +30,68 @@ pub struct UrlEncodedQuery;
...
+    EmptyQuery
+}
+
 /// Hashmap mapping strings to vectors of strings.
 pub type QueryMap = HashMap<String, Vec<String>>;
 
-impl Assoc<QueryMap> for UrlEncodedQuery {}
-impl Assoc<QueryMap> for UrlEncodedBody {}
+impl Pluggable for UrlEncodedQuery {}
+
```
Comment (reem) [inline src/lib.rs]: no newline needed here

File: src/lib.rs [hunk 7]
```diff
@@ -33,48 +30,68 @@ pub struct UrlEncodedQuery;
...
+    let mut deduplicated: QueryMap = HashMap::new();
 
     for (k, v) in q.into_iter() {
-        match deduplicated.entry(k) {
+        let is_new = match deduplicated.entry(k.clone()) {
             // Already a Vec here, push onto it
-            Occupied(entry) => { entry.into_mut().push(v); },
-
+            Entry::Occupied(entry) => { entry.into_mut().push(v.clone()); false},
+            Entry::Vacant(_) => true
```
Comment (reem) [inline src/lib.rs]: This refactoring confuses me. Why can't we do the insertion with the vacant entry?

Comment (hugoduncan) [inline src/lib.rs]: Confuses me too :). [Entry](http://doc.rust-lang.org/nightly/std/collections/hash_map/enum.Entry.html) no longer seems to have a `set` function.  There has to be a better way than what I have so far though.

Comment (reem) [inline src/lib.rs]: It's on the `VacantEntry` type as `insert`, which is found inside of the `Vacant` variant of `Entry`.

File: src/lib.rs (hunk 4 above)
Comment (hugoduncan) [inline src/lib.rs]: Sorry, you'll have to be more explicit.  What (and where) is the body reader plugin?

**Comment by hugoduncan** [comment]
I'm wondering if EmptyQuery should be returned as an error.  In the case where the query is optional, having an error, rather than an empty map, makes handling this case in user code a little complex and, I think, prevents the use of `get_ref`.

**Comment by s-panferov** [comment]
@hugoduncan iron/body-parser#43 has landed, so you can use it here now.

**Comment by reem** [comment]
whoops, sorry.

**Comment by hugoduncan** [comment]
I've written the body handling using map functions.  Let me know if you would prefer a `match` for handling the `Raw` result.

**Comment by reem** [comment]
Superseded by #47

**Pull request closed without merging**